Forge.TreeWalker/src/TreeWalkerSession.cs

```diff
@@ -654,7 +655,8 @@ internal async Task PerformActionTypeBehavior(TreeNode treeNode, string treeNode
 
             // Attmpt to ExecuteAction based on RetryPolicy and Timeout.
             // Throw on non-retriable exceptions.
-            while (actionTimeout == -1 || stopwatch.ElapsedMilliseconds < actionTimeout)
+            while (((actionTimeout == -1 || stopwatch.ElapsedMilliseconds < actionTimeout) && (retryPolicyType != RetryPolicyType.FixedCount)) || 
+                    (retryPolicyType == RetryPolicyType.FixedCount && maxRetryCount > 0))
```

Comment:
We want to retry only if maxRetryCount is satisfied AND we haven't hit a timeout. If either of those are false, we want to break out. (FixedCount can use Timeout as well if they desire.)

while (   (retryPolicyType != RetryPolicyType.FixedCount || (retryPolicyType == RetryPolicyType.FixedCount && maxRetryCount > 0))
       && (actionTimeout == -1 || stopwatch.ElapsedMilliseconds < actionTimeout))

Reply:
Please place the "&&" on the second line instead of at the end of first line. I'll ping you offline how I want this to look since spacing is bad in codeflow.

---
In reply to: [401964883](https://github.com/microsoft/Forge/pull/13#discussion_r401964883) [](ancestors = 401964883)